src/librustdoc/html/render.rs

```diff
@@ -2234,7 +2234,46 @@ fn item_trait(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item,
             }
         }
 
-        for implementor in implementors {
+        let (local, foreign) = implementors.iter()
+            .partition::<Vec<_>, _>(|i| i.impl_.for_.def_id().map_or(true, |d| cache.paths.contains_key(&d)));
+
+        if !foreign.is_empty() {
+            write!(w, "
+                <h2 id='foreign-impls' class='section-header'>
+                  Implementations on Foreign Types<a href='#foreign-impls' class='anchor'></a>
+                </h2>
+            ")?;
+
+            for implementor in foreign {
+                // need to get from a clean::Impl to a clean::Item so i can use render_impl
+                if let Some(t_did) = implementor.impl_.for_.def_id() {
```

Comment:
alternatively, i think the body of this block could be something like:

```rust
if let Some(impl_item) = impls
    .find(|i| i.impl_item.def_id == implementor.def_id)
    .next()
{
    let i = &implementor.impl_;
    let impl_ = Impl { impl_item: impl_item.clone() };
    let assoc_link = AssocItemLink::GotoSource(impl_item.def_id,
    &i.provided_trait_methods);
    render_impl(w, cx, &impl_, assoc_link,
    RenderMode::Normal, impl_item.stable_since(), false)?;
}
```

or